Libraries/LibGC/Heap+Swift.swift

```diff
@@ -0,0 +1,105 @@
+/*
+ * Copyright (c) 2024, Andrew Kaster <andrew@ladybird.org>
+ *
+ * SPDX-License-Identifier: BSD-2-Clause
+ */
+
+import AK
+@_exported import GCCxx
+
+extension GC.Heap {
+    public func withDeferredGC<R, E>(_ body: () throws(E) -> R) throws(E) -> R {
+        self.defer_gc()
+        defer { self.undefer_gc() }
+        return try body()
+    }
+}
+
+// FIXME: Cell and Cell::Visitor are not imported properly, so we have to treat them as OpaquePointer
+public protocol HeapAllocatable {
+    static func allocate(on heap: GC.Heap) -> UnsafeMutablePointer<Self>
+
+    init(cell: OpaquePointer)
+
+    func finalize()
+    func visitEdges(_ visitor: OpaquePointer)
+
+    var cell: OpaquePointer { get }
+}
+
+// FIXME: Figure out why other modules can't conform to HeapAllocatable
+public struct HeapString: HeapAllocatable {
+    public var string: Swift.String
+
+    public init(cell: OpaquePointer) {
+        string = "Hello, World!"
```

Comment:
That's cute, but maybe some like (untested):
```swift
public protocol HeapAllocatable {
    associatedtype Parameters
    init(cell: OpaquePointer, _ params: Parameters)
   ...
}

public struct HeapString: HeapAllocatable {
    public var value: Swift.String
    // type of params could _probably_ be extracted with a macro :P
    public init(cell: OpaquePointer, _ params: (string: Swift.String)) {
        self.value.init(params)
        self.cell = cell
    }
    ...
}
```
could work?

Reply:
I'm not sure that's super useful. For each type, we can just define another init() that takes those extra args. The only thing that the HeapAllocate protocol needs is for you to stash your ForeignCell pointer for folks to visit()

Reply:
We would also need a way to pass those values through some FFI hoops, as the HeapAllocatable.init(cell:) method is called via the Vtable's initialize pointer.

I think we'll need a different way to do complex init for these types. Either a two-phase init or some other runtime magic.,

Reply:
But allowing this init() means forcing everything to have a default constructor, that doesn't seem meaningful to me

Reply:
Right now we initialize the swift object via the vtable's inititialize() to call initializeFromFFI.

That method is like so:

```
    fileprivate static func initializeFromFFI(at this: UnsafeMutableRawPointer, cell: OpaquePointer) {
        this.assumingMemoryBound(to: Self.self).initialize(to: Self.self.init(cell: cell))
    }
 ```
 
 This method is called from ForeignCell::ForeignCell().
 
 We could maybe(?) end up calling a different per-type init method if we force a two-phase init on the ForeignCell instead, but that would leave the state where we have a valid Cell with an invalid trailing object visible to C++. 
 
A HeapAllocatable object could also create its own allocate(on:) method if it wants. 
 
 I think for the first cut requiring a default ctor for data members of HeapAllocatable objects is fine. We can work on adding more runtime sauce to do the two-phase init and initialize the rest of the data members with meaningful values as a follow-up. At that point we'll have a lot of manual machinery in place and it might be a good starting point for some macros to automate it.